gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-oidc/src/main/java/io/gravitee/am/gateway/handler/oauth2/service/granter/code/AuthorizationCodeTokenGranter.java

```diff
@@ -123,9 +129,16 @@ protected Single<TokenRequest> parseRequest(TokenRequest tokenRequest, Client cl
                                         // {#context.attributes['authFlow']['entry']}
                                         tokenRequest1.getContext().put(ConstantKeys.AUTH_FLOW_CONTEXT_ATTRIBUTES_KEY, ctx.getData());
 
-                                        // store the resource so it can be used later in the token creation for aud
-                                        tokenRequest1.setResources(authorizationCode.getResources());
-                                        return tokenRequest1;
+                                        // Resolve and validate final resources according to RFC 8707
+                                        return Single.fromCallable(() -> {
+                                                    // Preserve original authorization resources in a dedicated field
+                                                    tokenRequest1.setOriginalAuthorizationResources(authorizationCode.getResources());
+                                                    logger.debug("Preserved original authorization resources: {}", authorizationCode.getResources());
+
+                                                    Set<String> finalResources = resourceConsistencyValidationService.resolveFinalResources(tokenRequest1, authorizationCode.getResources());
+                                                    tokenRequest1.setResources(finalResources);
+                                                    return tokenRequest1;
+                                                }).toMaybe();
```

Comment:
from what I can see doing a fromCallable is useless here.

IMO we can remove  `return Single.fromCallable(() -> { ` and ` }).toMaybe();` and revert the `flatMap` to a `map`